File: server/internal/models/user_preferences.go
package models

import (
	"github.com/google/uuid"
	"github.com/juevigrace/diva-server/internal/models/responses"
	"github.com/juevigrace/diva-server/storage"
)

type UserPreferences struct {
	ID                  uuid.UUID
	UserID              uuid.UUID
	Theme               Theme
	OnboardingCompleted bool
	Language            string
	LastSyncAt          int64
	CreatedAt           int64
	UpdatedAt           int64
}

func (up *UserPreferences) Response() *responses.UserPreferencesResponse {
	return &responses.UserPreferencesResponse{
		Id:                  up.ID.String(),
		Theme:               up.Theme.String(),
		OnboardingCompleted: up.OnboardingCompleted,
		Language:            up.Language,
		LastSyncAt:          up.LastSyncAt,
		CreatedAt:           up.CreatedAt,
		UpdatedAt:           up.UpdatedAt,
	}
}

func (up *UserPreferences) DBCreate(userID uuid.UUID) *storage.CreateUserPreferencesParams {
	return &storage.CreateUserPreferencesParams{
		ID:                  up.ID,
		UserID:              userID,
		Theme:               up.Theme.ToDB(),
		OnboardingCompleted: up.OnboardingCompleted,
		Language:            up.Language,
	}
}

func (up *UserPreferences) DBUpdate() *storage.UpdateUserPreferencesParams {
	return &storage.UpdateUserPreferencesParams{
		ID:       up.ID,
		Theme:    up.Theme.ToDB(),
		Language: up.Language,
	}
}

func UserPrefsFromDB(row *storage.DivaUserPreference) *UserPreferences {
	return &UserPreferences{
		ID:                  row.ID,
		UserID:              row.UserID,
		Theme:               ThemeFromDB(row.Theme),
		OnboardingCompleted: row.OnboardingCompleted,
		Language:            row.Language,
		LastSyncAt:          row.LastSyncAt,
		CreatedAt:           row.CreatedAt,
		UpdatedAt:           row.UpdatedAt,
	}
}

File: kmp/app/core/models/models-core/src/commonMain/kotlin/com/diva/models/user/preferences/UserPreferences.kt
package com.diva.models.user.preferences

import com.diva.models.Theme
import com.diva.models.api.user.preferences.responses.UserPreferencesResponse
import com.diva.models.safeValueOfTheme
import io.github.juevigrace.diva.core.Option
import kotlin.time.Clock
import kotlin.time.Instant
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
data class UserPreferences(
    val id: Uuid = Uuid.NIL,
    val userId: Uuid = Uuid.NIL,
    val theme: Theme = Theme.SYSTEM,
    val onboardingCompleted: Boolean = false,
    val language: String = "en",
    val lastSyncAt: Option<Instant> = Option.None,
    val createdAt: Option<Instant> = Option.None,
    val updatedAt: Option<Instant> = Option.None,
) {
    companion object {
        fun fromResponse(response: UserPreferencesResponse): UserPreferences {
            return UserPreferences(
                id = Uuid.parse(response.id),
                theme = safeValueOfTheme(response.theme),
                onboardingCompleted = response.onboardingCompleted,
                language = response.language,
                lastSyncAt = Option.of(Instant.fromEpochMilliseconds(response.lastSyncAt)),
                createdAt = Option.of(Instant.fromEpochMilliseconds(response.createdAt)),
                updatedAt = Option.of(Instant.fromEpochMilliseconds(response.updatedAt))
            )
        }
    }
}
